pkg/util/signal.go

```diff
@@ -7,32 +7,38 @@ import (
 	"sync"
 )
 
-func NewSingleSignalPair() (SignalSender, SignalReceiver) {
-	sigCh := make(chan struct{})
+func NewSingleSignalPair[T any]() (SignalSender[T], SignalReceiver[T]) {
+	sigCh := make(chan T, 1)
 	once := &sync.Once{}
 	closeSigCh := func() { once.Do(func() { close(sigCh) }) }
-
-	return SignalSender{send: closeSigCh}, SignalReceiver{sigCh: sigCh, closeSigCh: closeSigCh}
+	sendOnce := &sync.Once{}
+
+	return SignalSender[T]{
+		send: func(data T) {
+			sendOnce.Do(func() { sigCh <- data })
```

Comment:
Having a separate `sendOnce` I don't think fully solves the panic issue, e.g.:

```go
tx, rx := NewSingleSignalPair()
rx.Close()
tx.Send(v)
```

will result in sending on the channel after it's already closed, right?

Consider having a single `sync.Once`, and using it for both the "close" and "send+close" operations?

Reply:
Ahh! The same once could potentially execute different things, I see.